Social Links: Add filter to services data to enable extendibility

mkaz opened the pull request:
## Description

Enable the ability to extend Social Links.

Adds a PHP filter to social links services to add a definition for a link and SVG.

Fixes #19041


## How has this been tested?

Create a plugin with that adds a filter to extend the services list, defining an icon and SVG for render side
Add JS that extends the variant and adds the icon for editor side.


### Example

PHP (tester.php)

```php
<?php
/*
        Plugin Name: Tester Plugin
        Description: A plugin for testing
*/

// Add admin JS file
add_action( 'enqueue_block_editor_assets',  function() {


        wp_enqueue_script(
                'tester-js',
                plugins_url( 'tester.js', __FILE__ ),
                array( 'wp-editor', 'wp-blocks', 'wp-dom-ready', 'wp-element' ),
                filemtime( plugin_dir_path( __FILE__ ) . 'tester.js' ), // cache buster
                true // in footer
        );

});

add_filter( 'block_editor_social_link_services', function( $data ) {
        $data[ 'mkaz' ] = array(
                'name' => 'mkaz',
                'icon' => '<svg width="24" height="24" viewBox="0 0 32 32" version="1.1" xmlns="http://www.w3.org/2000/svg" role="img" aria-hidden="true" focusable="false"><path d="M27 14h5c0-1.105-1.119-2-2.5-2s-2.5 0.895-2.5 2v0zM27 14c0-1.105-1.119-2-2.5-2s-2.5 0.895-2.5 2c0-1.105-1.119-2-2.5-2s-2.5 0.895-2.5 2v0 14c0 1.112-0.895 2-2 2-1.112 0-2-0.896-2-2.001v-1.494c0-0.291 0.224-0.505 0.5-0.505 0.268 0 0.5 0.226 0.5 0.505v1.505c0 0.547 0.444 0.991 1 0.991 0.552 0 1-0.451 1-0.991v-14.009c0-1.105-1.119-2-2.5-2s-2.5 0.895-2.5 2c0-1.105-1.119-2-2.5-2s-2.5 0.895-2.5 2c0-1.105-1.119-2-2.5-2s-2.5 0.895-2.5 2c0-5.415 6.671-9.825 15-9.995v-1.506c0-0.283 0.224-0.499 0.5-0.499 0.268 0 0.5 0.224 0.5 0.499v1.506c8.329 0.17 15 4.58 15 9.995h-5z"></path></svg>',
        );
        return $data;
} );
```

and JS (tester.js)

```js
( function( wp ) {

        wp.domReady(function() {

                const myIcon = wp.element.createElement(
                        'svg',
                        {
                                'viewBox': '0 0 32 32',
                                'width': '32',
                                'height': '32',
                        },
                        wp.element.createElement(
                                'path',
                                {
                                        'd': "M27 14h5c0-1.105-1.119-2-2.5-2s-2.5 0.895-2.5 2v0zM27 14c0-1.105-1.119-2-2.5-2s-2.5 0.895-2.5 2c0-1.105-1.119-2-2.5-2s-2.5 0.895-2.5 2v0 14c0 1.112-0.895 2-2 2-1.112 0-2-0.896-2-2.001v-1.494c0-0.291 0.224-0.505 0.5-0.505 0.268 0 0.5 0.226 0.5 0.505v1.505c0 0.547 0.444 0.991 1 0.991 0.552 0 1-0.451 1-0.991v-14.009c0-1.105-1.119-2-2.5-2s-2.5 0.895-2.5 2c0-1.105-1.119-2-2.5-2s-2.5 0.895-2.5 2c0-1.105-1.119-2-2.5-2s-2.5 0.895-2.5 2c0-5.415 6.671-9.825 15-9.995v-1.506c0-0.283 0.224-0.499 0.5-0.499 0.268 0 0.5 0.224 0.5 0.499v1.506c8.329 0.17 15 4.58 15 9.995h-5z",
                                }
                        )
                );

                wp.blocks.registerBlockVariation( 'core/social-link',
                        {
                                name: 'mkaz',
                                title: 'mkaz',
                                description: 'Test icon',
                                icon: () => myIcon,
                                attributes: { service: 'mkaz' },
                } );


                wp.blocks.unregisterBlockVariation( 'core/social-link', 'amazon' );

        } );


} )( window.wp );
```

## Types of changes

- Add a PHP filter, so an icon can be added on dynamic render.
- This required updating fetching icon and site name to pull from store and not hard-coded variant file

youknowriad commented:
In a previous PR we added a "templateLock" attribute to the "column" block to be able to edit it there.  In a lot of ways "templateLock" is similar to "allowedBlocks" and "template". An attribute allows templates to override these things for specific instances which is arguably more powerful.

I'm personally not at ease with using filters here (or any js filter for the record), it breaks too often depending on how/when scripts are loaded, components rerendered...

mkaz commented:
@youknowriad ~The issue I see using an attribute is it requires the developer to add or extend that component. For example, if I wanted to create my own `MyNavigationBlock` that extends the core block and allows for an additional `MyBlock` I could; however that adds complexity, plus for themes using a core navigation block it would not include my new navigation block.~

~So what I think developers really want is to just a way to add their `MyBlock` to the core allowed blocks list, which is why filters seems like the right approach, but I understand your hesitance on component rendering and loading, it isn't  consistent.~

I'll test out as an attribute, because then a plugin developer can override the setting using a register filter and achieve a similar result.

mkaz commented:
Updated, to use an attribute. I also updated the PR and testing instructions to match.

raaaahman commented:
Using attributes to list the allowed blocks makes sense to me. And it is filterable through the existing `blocks.registerBlockType` filter, which is good for extensibility by third-party plugins.

However, how would you proceed when the list is dynamically generated like in `social-links/edit.native.js` ? https://github.com/WordPress/gutenberg/blob/2190665a89f203a3e55409c03355b2f2a01a8920/packages/block-library/src/social-links/edit.native.js#L22-L24

Does it need to be dynamically generated in the first place? It is not obvious to me why the `edit.js` script allow the `core/social-link` block, while the `edit.native.js` needs the full list of social links variations.

gziolo commented:
I like the refined proposal after the feedback shared by @youknowriad. It's possible to change the list of default allowed attributes through `registerBlockType` filter when registering a block, or use `core/block-editor` store to set a different list of blocks by updating the value of the block attribute inside the `edit` implementation or externally through the data store.

fluiddot commented on packages/block-library/src/social-links/block.json:
```diff
@@ -3,6 +3,10 @@
 	"name": "core/social-links",
 	"category": "widgets",
 	"attributes": {
+		"allowedBlocks": {
+			"type": "array",
+			"default": [ "core/social-link" ]
+		},
```
I was wondering if this might be redundant since the block `core/social-link` will be always allowed because [its parent is this block](https://github.com/WordPress/gutenberg/blob/67de26f44e93e6274b822bc79e4b36dbe2588576/packages/block-library/src/social-link/block.json#L5). In fact, I tried to set the `allowedBlocks` to an empty array and I could still see all the social link variants displayed in the inserter.

gziolo replied:
Good catch. I completely forgot about it. It would be great to investigate how those two features interact with each other. The biggest limitation of `parent` is that it won't work with more than one level of nesting.

raaaahman replied:
So I tried fiddling with the `parent` property, and so far I am able to insert a custom block from a third-party plugin just by setting it a `parent` attribute.

Since child blocks are only available through their parent blocks, we could simply register two blocks that uses the same components for a block to be both available as an inner block and as a 'standalone' block.

For the nesting level concern, may it be something we want? Let's say third-plugin A adds its `a/social-share-link` in `core/social-link` inner blocks, while third-party plugin B adds its own `b/social-profile-avatar-display` which accepts an inner `core/social-link` block, and retrieves the profile picture from the passed social link (throughout some CORS bypass black magic). 

Wouldn't we, in this case, want to prevent the `a/social-share-link` to be available as an inner block of `b/social-profile-avatar-display`?

fluiddot replied:
>The biggest limitation of parent is that it won't work with more than one level of nesting.

Yeah, good point. Are the allowed blocks list passed through `allowedBlocks` prop propagated to nested children? I haven't checked this further but it looks like it doesn't neither, right?

mkaz replied:
This was a great catch because I was testing Navigation with the same update and trying to limit the list of blocks and hadn't figured out yet why it was still showing the other blocks.

mcsf commented:
👋 

I'm not sure how this fixes the parent issue, #19041. Is the goal to be able to insert third-party icon-like blocks alongside `core/social-link` children? If so, it seems to me like a missed opportunity to look at how `core/social-link` itself can be leveraged by third parties. Since that type is already taking advantage of the block variations mechanism, which is inherently flexible, I'd look there first.

Allowing new child types inside Social Links would lead to a more fragmented user experience, where some of the icons no longer behave identically — the toolbar would shift, as would the block settings UI. Also, we would miss out on Gutenberg's multi-select feature that lets the user set attributes for multiple blocks of the same type.

fluiddot commented:
> Using attributes to list the allowed blocks makes sense to me. And it is filterable through the existing `blocks.registerBlockType` filter, which is good for extensibility by third-party plugins.
> 
> However, how would you proceed when the list is dynamically generated like in `social-links/edit.native.js` ?
> 
> https://github.com/WordPress/gutenberg/blob/2190665a89f203a3e55409c03355b2f2a01a8920/packages/block-library/src/social-links/edit.native.js#L22-L24
> 
> Does it need to be dynamically generated in the first place? It is not obvious to me why the `edit.js` script allow the `core/social-link` block, while the `edit.native.js` needs the full list of social links variations.

Thanks @raaaahman for pointing this out!

I think we don't really need to dynamically generate them, I'm not sure what was the purpose of doing it this way, I checked the [PR](https://github.com/WordPress/gutenberg/pull/23017) that introduced it but I didn't get any clue regarding this.

I ran some tests and I found out that it doesn't look like it's necessary to pass the social link variants in the `allowedBlocks`, the reason is that these variants are registered using the `core/social-link` settings which defines `core/social-links` as its parent. This means that by default they will always be allowed and displayed in the inserter for the `core/social-links` block.

Here I reference some code pieces related to this:
- [Selector that determines if a block should be included in the inserter](https://github.com/WordPress/gutenberg/blob/eb957e77afbb68f6785bc17720b42a83f80142db/packages/block-editor/src/store/selectors.js#L1339-L1360)
- [Social link variations are registered as child blocks of `core/social-links` in the native version](https://github.com/WordPress/gutenberg/blob/82ddc9e405d91066eba6d08fb716ef8feea922ab/packages/block-library/src/index.native.js#L255)
- [In the native version the block `core/social-link` is set to not support inserter](https://github.com/WordPress/gutenberg/blob/82ddc9e405d91066eba6d08fb716ef8feea922ab/packages/block-library/src/index.native.js#L170-L179)

**EDIT:** Unless there's any potential issue I might be overlooking, I think we could apply the same logic in the native version of the `core/social-links` component.

geriux commented:
> I think we don't really need to dynamically generate them, I'm not sure what was the purpose of doing it this way, I checked the [PR](https://github.com/WordPress/gutenberg/pull/23017) that introduced it but I didn't get any clue regarding this.

I can't remember exactly why this decision was made but I believe it was mostly UX decisions, some things that come to mind:

- We probably wanted to sort them by title hence the `sortBy` to simplify the search of the desired block, we didn't have the search box back then but now that we do maybe we don't need to sort them.
- I remember we also had some issues with the floating toolbar to show some specific data icon/title:
<kbd><img width="250" alt="Screenshot 2021-04-15 at 13 24 36" src="https://user-images.githubusercontent.com/4885740/114862234-8f7c5700-9dee-11eb-86e7-ea7d35882219.png"></kbd>

I'm sure we could solve this in another way if we want to remove this custom `registerBlockVariations`.

fluiddot commented:
> > I think we don't really need to dynamically generate them, I'm not sure what was the purpose of doing it this way, I checked the [PR](https://github.com/WordPress/gutenberg/pull/23017) that introduced it but I didn't get any clue regarding this.
> 
> I can't remember exactly why this decision was made but I believe it was mostly UX decisions, some things that come to mind:
> 
> * We probably wanted to sort them by title hence the `sortBy` to simplify the search of the desired block, we didn't have the search box back then but now that we do maybe we don't need to sort them.
> * I remember we also had some issues with the floating toolbar to show some specific data icon/title:
>   <img alt="Screenshot 2021-04-15 at 13 24 36" width="250" src="https://user-images.githubusercontent.com/4885740/114862234-8f7c5700-9dee-11eb-86e7-ea7d35882219.png">
> 
> I'm sure we could solve this in another way if we want to remove this custom `registerBlockVariations`.

Thanks @geriux for the insights!

That was my impression too, it looks like the way the native version is registering the block variants is related to what you commented.
 
For this case what I suggest is only to modify the way the allowed blocks list is calculated in the `core/social-links`  block for passing it to the `InnerBlock` component. I don't think we would need to modify how the variations are registered, just to match the logic with the web version of this component regarding this part.

<details><summary><strong>Suggested change</strong></summary>

```
diff --git a/packages/block-library/src/social-links/edit.native.js b/packages/block-library/src/social-links/edit.native.js
index b52be29e70382c536245422e0a8160d38611f8ca..52d8b61934bfb7ae0b3cfe9cb13e7ab406d985dc 100644
--- a/packages/block-library/src/social-links/edit.native.js
+++ b/packages/block-library/src/social-links/edit.native.js
@@ -17,11 +17,6 @@ import { compose, usePreferredColorSchemeStyle } from '@wordpress/compose';
  * Internal dependencies
  */
 import styles from './editor.scss';
-import variations from '../social-link/variations';
-
-const ALLOWED_BLOCKS = variations.map(
-	( v ) => `core/social-link-${ v.name }`
-);
 
 // Template contains the links that show when start.
 const TEMPLATE = [
@@ -47,7 +42,7 @@ function SocialLinksEdit( {
 } ) {
 	const [ initialCreation, setInitialCreation ] = useState( true );
 	const shouldRenderFooterAppender = isSelected || isInnerIconSelected;
-	const { align } = attributes;
+	const { align, allowedBlocks } = attributes;
 	const { marginLeft: spacing } = styles.spacing;
 
 	useEffect( () => {
@@ -91,7 +86,7 @@ function SocialLinksEdit( {
 
 	return (
 		<InnerBlocks
-			allowedBlocks={ ALLOWED_BLOCKS }
+			allowedBlocks={ allowedBlocks }
 			templateLock={ false }
 			template={ initialCreation && TEMPLATE }
 			renderFooterAppender={
```

</details>

One thing I'd like to note is that the `core/social-link` block is explicitly disabled for the inserter in the native version, this shouldn't be a problem because the variants are still added via the parent setting.

mkaz commented:
@mcsf I think you are right, for Social we probably can just use the filter and extend the variations settings without needing a new attribute. The goal is to include the same extensibility to Navigation block which that would not work, and Navigation does make sense to allow and limit different block types.

However, as @fluiddot points out the child inherents from the parent, so limiting might be more difficult will need some more testing around that for Navigation.

Also, I'll try on Social, maybe its just a matter of documenting as a plugin developer how you would create your own variant to add an additional icon.

mcsf commented:
> as fluiddot points out the child inherents from the parent, so limiting might be more difficult will need some more testing around that for Navigation.

In this scenario, who is the parent and who is the child?

***

Looking at the quirks in the `native.js` patch, especially the parts that rest on the deprecated (as of #19887) implementation of Social Links as distinct block types (`variations.map( v => 'core/social-link-' + v.name )`), and looking at the discussion about [the `parent` property taking precedence over `allowedBlocks`](https://github.com/WordPress/gutenberg/pull/30749#discussion_r613865473), I can't help but think that we're hacking around the real issue, which is managing (extending, disabling) variations.

There is only one child type: `core/social-link`. It doesn't make sense to me to want to alter its `parent` property, because that type only makes sense within the parent `core/social-links`. I consider that a code smell. We either want new icons to be available, or fewer. On the client, this is what `registerBlockVariation` and `unregisterBlockVariation` ([doc](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-variations.md)) are meant to do. If they are proving insufficient, let's see what we can do there.

In the specific case of Social Links, we have the server-side counterpart to address. For this, I'd expect to be able to use [WP's rendering hooks](https://make.wordpress.org/core/2021/02/18/wordpress-5-7-a-new-dynamic-hook-to-filter-the-content-of-a-single-block/) to enhance the back-end with third-party variations.

***

Am I off base here?

mkaz commented:
@mcsf I don't think you're off base at all, precisely for Social Links. I was thinking of the implementation for allowedBlocks in Navigation and brought that mindset back to Social Links first, since the two are similar in some regards. I started off wrong and didn't think of the variations. I think we can probably just document and use that for Social, I agree additional block types does not make sense.

For extending Navigation, we'll need to do something else, and we can look at that in a separate PR as I explore that one more.

This PR for Social will likely just change to a documentation update on how to use variants to extend, hopefully I'll get to updating it today or tomorrow.

mkaz commented:
@mcsf After some testing and trying out variations it does work to some degree. The `unregisterBlockVariation` works nicely for removing any icons you don't want. 

The issue comes in with adding an icon, you can add an additional variant, and it shows up and can be selected without a problem The issue is on display, the icons are inserted using a dynamic render, the render block function will add the proper SVG hard-coded in the social links `index.php`. 

So I think the solution could be to add a filter on the PHP side to extend that list. Thoughts?

mkaz committed:
Add filter for social link services data

mkaz commented:
Ok, I tried an update adding a filter on the PHP side to add a service, and a filter on the JS side to add a variant.
I then created a plugin to add my variant and icon, (PR updated with test plugin code) and it gets us oh so close!

See the video, the one last spot would be the icon in the block toolbar does not get replaced:


https://user-images.githubusercontent.com/45363/115069438-68825b80-9ea8-11eb-9e11-a54034baa98d.mp4

mkaz committed:
Retrieve variant list from store, in case extended

gziolo commented on packages/block-library/src/social-link/edit.native.js:
```diff
@@ -65,8 +65,8 @@ const SocialLinkEdit = ( {
 
-	const IconComponent = getIconBySite( service )();
-	const socialLinkName = getNameBySite( service );
+	const IconComponent = GetIconBySite( service )();
```
React hooks like `useSelect` won't work as expected this way, they need to be rendered as components.

gziolo replied:
Actually, those could be React hooks:

```js
const socialLinkIcon = useServiceIcon( service );
const socialLinkName = useServiceName( service );
```

mkaz replied:
I fixed this in 0edca09 by removing the `()` and using the `IconComponent` directly in render

gziolo replied:
React docs discourage nesting hooks in regular functions:

https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level

> **Don’t call Hooks inside loops, conditions, or nested functions.** Instead, always use Hooks at the top level of your React function, before any early returns. By following this rule, you ensure that Hooks are called in the same order each time a component renders.

I'm not sure why they advocate for that but it as simple as replacing `Get` with `use` in function names.

mcsf replied:
> I'm not sure why they advocate for that

Abramov's _[Why do hooks rely on call order?](https://overreacted.io/why-do-hooks-rely-on-call-order/)_ is a great starting point. Some of his other posts ([1](https://overreacted.io/algebraic-effects-for-the-rest-of-us/), [2](https://overreacted.io/why-isnt-x-a-hook/)) provide deep complementary insight.

mkaz replied:
@gziolo I switched the two functions to hooks :+1:

jasmussen commented:
This is really cool. I won't speak to the code, as it seems you've already been getting a bunch of great feedback there. Here's what I saw:

![ey](https://user-images.githubusercontent.com/1204802/115356231-bb675600-a1bb-11eb-88b8-05ecb3ad7a6e.gif)

Here's the JS error:

<img width="761" alt="Screenshot 2021-04-20 at 09 34 51" src="https://user-images.githubusercontent.com/1204802/115356262-c1f5cd80-a1bb-11eb-9ebd-8a5a6cf6cce3.png">

Note that I did not follow the instructions perfectly, in fact I just dumped the contents of your suggested plugin in my functions.php assuming it'd work the same. It might not have?

For one, in order to load the JS file I also had to tinker a bit.

`plugins_url( 'tester.js', __FILE__ ); ` for me returned http://localhost:8888/wp-content/plugins/var/www/html/wp-content/themes/mocco-dev/tester.js

So to get it to work I changed it to `get_template_directory_uri() . '/tester.js',`

I'm excited about this! Let me know what I did wrong and I'll give it another test.

mkaz committed:
Update native usage of IconComponent

The native implementation was () the component and then rendering
using the <Icon/> this does not work with hooks added.

This PR switches to using the IconComponent directly and passing the
additional animated and styles props to it.

mkaz commented:
@jasmussen I forgot to update my example, I made the same mistake the first time and only corrected it locally. 

The example above is now updated with the fix. The change is in the `tester.js` file, the icon property in the `registerBlockVariant` needs to be a functional component, this can be changed to this:

```js
       icon: () => myIcon,
```

jasmussen commented:
Got it working:

![hey](https://user-images.githubusercontent.com/1204802/115542760-e4ade200-a2a0-11eb-9a27-94a3e2954923.gif)

Don't mind the black button in this case, it's CSS from my own theme that I need to tinker with. I can push some polish here!

mcsf commented on packages/block-library/src/social-link/index.php:
```diff
@@ -278,6 +278,8 @@ function block_core_social_link_services( $service = '', $field = '' ) {
 		),
 	);
 
+	$services_data = apply_filters( 'block_editor_social_link_services', $services_data );
+
```
I'm not in love with:

* Adding a block-specific API in our core library
* The fact that this API exposes implementation details (`$services_data`) of that block's rendering logic

I'd love to find a solution that could inform other multi-variation server-rendered blocks, or at least a solution that is more consistent with how we already extend or intercept blocks. In the absence of that, I almost prefer that third parties just use the standard `render_block_core/social-link` filter to intercept a block that matches their variation and swap out the serialised SVG.

<details><summary>Example</summary><pre><code>
add_filter(
	'render_block_core/social-link',
	function ( $block_content, $parsed_block ) {
		if ( 'my-variation' === $parsed_block['attrs']['service'] ) {
			preg_replace( ... );
		}
		return $block_content;
	},
	10,
	2
);
</code></pre></details>

I'm not sure. What do you think, @gziolo?

gziolo replied:
Some thoughts I have on this topic.

First of all, I see that we use `block_core_social_link_services` to keep the list of all available services on the server and we keep a very similar list of services as an array of variations on the client. They are mostly the same but keep the same data in different formats. The biggest difference present is for icons that are React elements from `@wordpress/icons` on the client and regular SVGs on the server.

In https://github.com/WordPress/gutenberg/pull/29095, @gwwar added block variations for the Navigation Link block on the server. We had a few follow-up tasks that improve the integration. One missing bit is the support in `block.json` file for block variations which might not work at all here. My initial reaction was that we could explore if we could do all the filtering with PHP on the server with `register_block_type_args` (or even better we add `register_block_type_{$setting_name}`) to register new block variations. However it doesn't solve the issue with icons.

Knowing all that, it is quite interesting approach to make `block_core_social_link_services` extensible despite the fact that it introduces this level of indirection through another filter that has to be documented. It might be simpler to go this route than to teach folks how to replace some pattern in the `render_block` filter. 

I don't think I'm helping much here. I wanted also to emphasize that we need to improve support for block icons that has to be defined on the client since we started using `@wordpress/icons`. It was also raised by @gwwar recently in the context of the Navigation Link block in https://github.com/WordPress/gutenberg/issues/30739#issuecomment-817908133:

> Sidenote: To fully remove the hooks.js file from navigation-link, 🤔 thinking through icons would be another one. One straightforward approach would be passing a string matching the icon name from the package. Is there any background on why we may have avoided this?

mcsf replied:
Thanks for weighing in.

> Knowing all that, it is quite interesting approach to make `block_core_social_link_services` extensible despite the fact that it introduces this level of indirection through another filter that has to be documented. It might be simpler to go this route than to teach folks how to replace some pattern in the `render_block` filter.

As long as we're aware of the tradeoffs, that's good enough for me. @mkaz: if I recall correctly, back when Social Icons were introduced, we went with SVGs because Internet Explorer didn't support the alternative — is that correct? If so, what was the alternative? I ask this because that could inform our mid-term solutions for letting third parties provide new icons. In other words, once we [rid ourselves of the IE burden](https://make.wordpress.org/core/2021/03/25/discussion-summary-dropping-support-for-ie11/), can we come up with better APIs?

mkaz replied:
@jasmussen will have the details better. I know we've always wanted to use SVG because they scale nicely. The main issues we've run into is inserting SVG into post content (requires admin privileges) and then how to embed them using CSS is I believe the IE conflict.

Also, I'm not sure how much IE support we would want to eliminate when publishing content. Not requiring IE for using the editor and admin screens is one thing, not having things show up to visitors is another.

jasmussen replied:
It's a bit far back now, but I think we actually decided to treat IE(11) in a progressive-enhancement way. The conversation, if I recall correctly, was about easily recolored SVGs, and I think we found a way to show the same SVGs to IE11, just with reduced or no colors. 

SVG is still ideal to me, as they are easy to make, they are easy to colorize, and you can even do fancy animation stuff with them.

If we were to put on our dreamer-hats and relieve ourselves of even more burdens, we should look at improving caching of the SVGs by having one or more external files that are reused across pages, rather than outputting the entire SVG inline as we do now. All the while, the SVG still has to be easily recolored.

The obvious method for sprite work is SVG `use`, which is essentially that, and this article explains it well: https://css-tricks.com/svg-use-external-source/

The primary downside with that is that as an external sprite, extensibility might be tricky (or it might not? maybe extensons just don't point to a sprite, but to an SVG file?).

An alternative is to use `mask-image`, as demonstrated in this codepen: https://codepen.io/joen/pen/poRwpPj

The benefits are external files, easy recolorability, pretty simple code, and should work well for extensibility. The downside is browser support (though we should be [reasonably good](https://caniuse.com/css-masks) if we discount IE11), and the fact that you can only have single-color icons, because the icon gets "flattened" when used as a mask shape. That last tradeoff seems fine to me for social links, even if that does post a problem for the Snapchat logo:

<img width="112" alt="Screenshot 2021-04-22 at 19 07 05" src="https://user-images.githubusercontent.com/1204802/115755726-f2955d00-a39d-11eb-8e21-800ccb53e9fa.png">

mcsf replied:
Fair enough. Well, don't let me block you on this one anymore. :)

mcsf replied:
My previous comment was in response to Marcus; GitHub hadn't yet refreshed, so I just saw Joen's comment now.

Anyway, this is pretty cool:

> An alternative is to use mask-image, as demonstrated in this codepen: https://codepen.io/joen/pen/poRwpPj

mkaz committed:
Switch fetching icon/name to hooks

mkaz commented:
I believe this PR is ready to land.

Summarizing the two sub-conversations:

1. Around hooks and functions - the functions that fetch name and icon were switched to hooks, so should resolve those concerns.

2. Around IE, SVG and sprites - IE11 support is being dropped in WP 5.8, but this is for the admin and editor screens, I don't think we should alter markup output to visitors yet, especially since the result would likely be empty. So I think it is still too soon to adopt the mask or sprite features Joen described.

gwwar commented:
Sorry, I'm a bit late looking at this one, but this feels pretty specific to a single block and very close to the server icon variation problems I ran into with navigation-link, do you mind giving me a half-day to catch up for a review?

mkaz commented:
@gwwar No problem, no one's approved it yet anyways :smile: 

The PR was initially started as a means of trying to address extendability for both Social and Navigation, but it has meandered a little and ended a bit more specific to Social due to how icons are loaded. Ping me with any questions.

gwwar commented:
To recap a quick slack chat:

Currently social icons are dynamic blocks to avoid wp_kses issues with SVG + user roles. That is, if an author saves a post with social logos in it, we shouldn't strip out the SVGs from post_content. With this implementation choice, however we were forced to duplicate icon information in JS/PHP, without a way to extend this.

As @gziolo noted, we did start exploring server defined variations, but it was partially implemented due to variation.icon and variation.isActive being functions which cannot be serialized. In https://github.com/WordPress/gutenberg/pull/30913 isActive was updated to accept a function or string array. 

I think we could do a similar exploration with icon to accept a react component function, a mapped string name (like "home") from our [icon library](https://wordpress.github.io/gutenberg/?path=/story/icons-icon--library), or a string svg like `<svg width="24" height="24" viewBox="0 0 24 24" version="1.1" xmlns="http://www.w3.org/2000/svg" role="img" aria-hidden="true" focusable="false"><path d="M12.158,12.786L9.46,20.625c0.8...></svg>`.

I can spin up a draft wordpress-develop PR if folks are interested in seeing that. As a first step for example we might update WordPress to accept variation.icon as a string value, and see how that works together with the current social logos implementation.

gziolo commented:
> I think we could do a similar exploration with icon to accept a react component function, a mapped string name (like "home") from our [icon library](https://wordpress.github.io/gutenberg/?path=/story/icons-icon--library), or a string svg like `<svg width="24" height="24" viewBox="0 0 24 24" version="1.1" xmlns="http://www.w3.org/2000/svg" role="img" aria-hidden="true" focusable="false"><path d="M12.158,12.786L9.46,20.625c0.8...></svg>`.

One important factor to take into account is that it has to work also with React Native. The current setup where all icons are React components integrates nicely with `react-native-svg` library. I see there is an option to use XML strings that looks close to what you proposed but it was never tested with the mobile app:
https://github.com/react-native-svg/react-native-svg#use-with-xml-strings

You can also check `SVG` components to see how interoperability is implemented:
https://github.com/WordPress/gutenberg/tree/trunk/packages/primitives/src/svg

tomjn commented:
@mkaz I see new commits have been pushed up since the last notes, does this mean the SVG issue has been worked on and the PR is ready for evaluation/testing again? Or is more work necessary first?

mkaz commented:
@tomjn Thanks for the ping Tom, I've got this tab open and will follow up and make sure it's ready. I believe when I last left it was but can read through it all again and see if anything has changed.

BeholdPL commented:
Is code from above is still working? Unfortunately I have problem with proper icon - in elements browser looks good, but the rendered icon is wrong.

js:

```
(function(wp) {
  wp.domReady(function() {
    const messengerIcon = wp.element.createElement(
      'svg', {
        'viewBox': '0 0 512 512',
        'width': '512',
        'height': '512',
      },
      wp.element.createElement(
        'path', {
          'd': "M256.55 8C116.52 8 8 110.34 8 248.57c0 72.3 29.71 134.78 78.07 177.94 8.35 7.51 6.63 11.86 8.05 58.23A19.92 19.92 0 0 0 122 502.31c52.91-23.3 53.59-25.14 62.56-22.7C337.85 521.8 504 423.7 504 248.57 504 110.34 396.59 8 256.55 8zm149.24 185.13l-73 115.57a37.37 37.37 0 0 1-53.91 9.93l-58.08-43.47a15 15 0 0 0-18 0l-78.37 59.44c-10.46 7.93-24.16-4.6-17.11-15.67l73-115.57a37.36 37.36 0 0 1 53.91-9.93l58.06 43.46a15 15 0 0 0 18 0l78.41-59.38c10.44-7.98 24.14 4.54 17.09 15.62z",
        }
      )
    );
    wp.blocks.registerBlockVariation('core/social-link', {
      name: 'messenger',
      title: 'Facebook Messenger',
      description: 'Facebook Messenger',
      icon: () => messengerIcon,
      attributes: {
        className: 'ml-messenger',
        providerNameSlug: 'facebook-messenger',
        responsive: true
      },
    });
  });
})(window.wp);
```


php:

```
function messengerSocialMediaIcon() {
	wp_enqueue_script( 'social_media', get_template_directory_uri() . '/js/social_media.js', array( 'wp-editor', 'wp-blocks', 'wp-dom-ready', 'wp-element' ), true );
}
add_action( 'enqueue_block_editor_assets', 'messengerSocialMediaIcon' );

add_filter( 'block_editor_social_link_services', function( $data ) {
	$data[ 'messenger' ] = array(
			'name' => 'messenger',
			'icon' => '<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path fill="currentColor" d="M256.55 8C116.52 8 8 110.34 8 248.57c0 72.3 29.71 134.78 78.07 177.94 8.35 7.51 6.63 11.86 8.05 58.23A19.92 19.92 0 0 0 122 502.31c52.91-23.3 53.59-25.14 62.56-22.7C337.85 521.8 504 423.7 504 248.57 504 110.34 396.59 8 256.55 8zm149.24 185.13l-73 115.57a37.37 37.37 0 0 1-53.91 9.93l-58.08-43.47a15 15 0 0 0-18 0l-78.37 59.44c-10.46 7.93-24.16-4.6-17.11-15.67l73-115.57a37.36 37.36 0 0 1 53.91-9.93l58.06 43.46a15 15 0 0 0 18 0l78.41-59.38c10.44-7.98 24.14 4.54 17.09 15.62z"></path></svg>',
	);
	return $data;
} );
```

results:

![Screenshot 2021-11-13 at 23 21 24](https://user-images.githubusercontent.com/34622592/141661027-90c65d38-c920-4783-b11f-cb92bc6904ce.png)

Is `block_editor_social_link_services` filter working?

Tested at Wordpress v.5.8.2

tomjn commented:
@BeholdPL are you testing the code with this PR?

BeholdPL commented:
> @BeholdPL are you testing the code with this PR?

Tested on WP 5.8.2

tomjn commented:
> Tested on WP 5.8.2

Then there's the problem, the code in this pull request is intended for testing this pull request, not as an example for general use. _This PR has not been merged._

To use this code, you would need to pull down this branch of the gutenberg plugin, build it, and activate it. There is no guarantee it will work, or that the PR does not need additional work. It is not a copy pasta example for common use.

tomjn commented:
Any results you have gotten in 5.8.2 are accidental/coincidences. I know that the filter proposed here is in use in the twentytwentyone theme for use on the frontend though with slight differences

BeholdPL commented:
> Any results you have gotten in 5.8.2 are accidental/coincidences. I know that the filter proposed here is in use in the twentytwentyone theme for use on the frontend though with slight differences

Got it. As far as I know there is no way for extending this block in current WP version (5.8.2), isn't it?

mkkeck commented:
> Got it. As far as I know there is no way for extending this block in current WP version (5.8.2), isn't it?

You can try the Patched Block:

https://gist.github.com/mkkeck/84fe5fda939002fcf210d66b99de7c0f

NickHeurter commented:
> > Got it. As far as I know there is no way for extending this block in current WP version (5.8.2), isn't it?
> 
> You can try the Patched Block:
> 
> https://gist.github.com/mkkeck/84fe5fda939002fcf210d66b99de7c0f

@mkkeck How can I use this in my WordPress website without the need to be afraid for updated? :)

mkkeck commented:
Use it simply as a plugin.
I'm using this since WP 5.6 and it Work with current WP version.

NickHeurter commented:
> 

I'm sorry to ask @mkkeck, but I can't get it working. Could you share your full plugin with social-services.json file by any change? Sorry for asking!

mkkeck commented:
Please take look at lines #84 until #97:

```
$json_file = apply_filters(
   'PatchBlocks/SocialServicesJsonFile', 
  // Theme class to check if file exists and is readable
  // add here your own   
  Theme_Defaults::instance()->get_style_dir('social-services.json') 
  // or use 
  // get_stylesheet_directory() . 'social-services.json'
);
```

I think you have to replace it with:
```
$json_file = apply_filters(
   'PatchBlocks/SocialServicesJsonFile', 
  get_stylesheet_directory() . 'social-services.json'
);
```

sgeray commented:
> 

Thanks to Cameron Jones ( https://cameronjonesweb.com.au/blog/adding-a-custom-icon-to-the-social-links-block-should-not-be-this-hard/ ) I am able to add TripAdvisor icon as well.

```
add_filter(
	'render_block',
	function( $block_content, $block ) {
		if ( isset($block['attrs']['service']) && 'core/social-link' === $block['blockName'] && 'tripadvisor' === $block['attrs']['service']  ) {
			$icon          = '<svg viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg"><path d="m6.781 12.894v.016c0 .266-.11.505-.286.677-.17.177-.408.286-.672.286-.004 0-.007 0-.011 0h.001s-.001 0-.002 0c-.54 0-.978-.438-.978-.978 0-.001 0-.001 0-.002 0-.003 0-.007 0-.01 0-.264.11-.502.286-.672.172-.177.412-.287.677-.287h.017-.001c.533.005.963.436.969.968v.001zm12.019-.013c0 .541-.438.979-.979.979s-.979-.438-.979-.979c0-.005 0-.011 0-.017 0-.263.11-.501.286-.67.179-.174.423-.281.693-.281s.514.107.693.282c.177.169.287.407.287.671v.017-.001zm-10.832.01c-.001-1.109-.9-2.007-2.009-2.007s-2.009.899-2.009 2.009.899 2.009 2.009 2.009c.555 0 1.057-.225 1.421-.589.363-.364.587-.867.588-1.421zm12.009-.01c0-1.109-.9-2.008-2.009-2.008s-2.009.899-2.009 2.009.899 2.009 2.009 2.009c.555 0 1.057-.225 1.42-.588.363-.353.589-.846.589-1.391 0-.011 0-.021 0-.032v.002zm-11.125.01c0 1.602-1.299 2.901-2.901 2.901s-2.901-1.299-2.901-2.901c0-.012 0-.026 0-.041 0-.788.328-1.499.854-2.005l.001-.001c.51-.521 1.221-.844 2.007-.844h.048-.002.041c.785 0 1.494.324 2.001.846l.001.001c.524.508.85 1.218.85 2.005v.044-.002zm12.02-.01v.004c0 1.602-1.299 2.901-2.901 2.901s-2.901-1.299-2.901-2.901 1.299-2.901 2.901-2.901c.801 0 1.527.325 2.052.85.524.508.849 1.218.849 2.004v.045zm-10.041.031c0-.021.001-.045.001-.07 0-1.307-.541-2.487-1.41-3.33l-.001-.001c-.844-.871-2.024-1.412-3.331-1.412-.025 0-.049 0-.074.001h.004c-.01 0-.021 0-.032 0-.875 0-1.694.24-2.394.658l.021-.012c-1.443.85-2.397 2.397-2.397 4.166s.953 3.316 2.374 4.154l.022.012c.679.406 1.498.646 2.373.646h.036-.002c.021 0 .045.001.07.001 1.307 0 2.487-.541 3.33-1.41l.001-.001c.871-.844 1.412-2.024 1.412-3.331 0-.024 0-.049-.001-.073v.004zm6.96-5.969c-1.66-.729-3.594-1.154-5.628-1.154-.057 0-.115 0-.172.001h.009c-.071-.001-.154-.002-.238-.002-2.068 0-4.038.422-5.827 1.184l.097-.037h.017c.832 0 1.624.175 2.34.489l-.037-.015c.739.315 1.37.746 1.901 1.276.526.534.955 1.167 1.255 1.867l.015.039c.3.679.474 1.471.474 2.303v.018-.001c0-1.619.641-3.088 1.682-4.168l-.002.002c.519-.524 1.131-.955 1.811-1.264l.038-.016c.662-.311 1.436-.501 2.252-.521h.007zm4.99 5.968c-.003-2.653-2.154-4.802-4.807-4.802-2.655 0-4.807 2.152-4.807 4.807s2.152 4.807 4.807 4.807c1.329 0 2.531-.539 3.402-1.41.868-.844 1.406-2.023 1.406-3.327 0-.026 0-.053-.001-.079zm-2.77-5.895h3.989c-.298.346-.558.738-.766 1.16l-.014.032c-.181.344-.325.743-.412 1.163l-.005.03c.715.972 1.144 2.193 1.144 3.515 0 3.297-2.673 5.97-5.97 5.97-1.864 0-3.529-.855-4.624-2.194l-.008-.011c-.436.541-.879 1.15-1.291 1.781l-.054.087c-.191-.333-.376-.613-.576-.882l.016.023q-.442-.63-.776-1.015c-1.103 1.324-2.753 2.161-4.598 2.161-3.298 0-5.971-2.673-5.971-5.971 0-1.298.414-2.499 1.117-3.478l-.012.018c-.092-.45-.236-.849-.429-1.218l.012.025c-.223-.455-.484-.847-.789-1.2l.006.007h3.802c1.062-.701 2.293-1.26 3.61-1.611l.088-.02c1.327-.373 2.851-.588 4.425-.588h.079-.004.107c1.523 0 2.995.213 4.39.611l-.113-.027c1.378.383 2.584.942 3.672 1.665l-.047-.029z"/></svg>';
			$before        = explode( '<svg', $block_content );
			$after         = explode( '</svg>', $before[1] );
			$block_content = $before[0] . $icon . $after[1];
		}
		return $block_content;
	},
	10,
	2
);
```

and the block;

```
import { Path, SVG } from '@wordpress/primitives';

export const tripadvisorIcon = () => (
    <SVG viewBox="0 0 24 24" xmlns="http://www.w3.org/2000/svg">
        <Path d="m6.781 12.894v.016c0 .266-.11.505-.286.677-.17.177-.408.286-.672.286-.004 0-.007 0-.011 0h.001s-.001 0-.002 0c-.54 0-.978-.438-.978-.978 0-.001 0-.001 0-.002 0-.003 0-.007 0-.01 0-.264.11-.502.286-.672.172-.177.412-.287.677-.287h.017-.001c.533.005.963.436.969.968v.001zm12.019-.013c0 .541-.438.979-.979.979s-.979-.438-.979-.979c0-.005 0-.011 0-.017 0-.263.11-.501.286-.67.179-.174.423-.281.693-.281s.514.107.693.282c.177.169.287.407.287.671v.017-.001zm-10.832.01c-.001-1.109-.9-2.007-2.009-2.007s-2.009.899-2.009 2.009.899 2.009 2.009 2.009c.555 0 1.057-.225 1.421-.589.363-.364.587-.867.588-1.421zm12.009-.01c0-1.109-.9-2.008-2.009-2.008s-2.009.899-2.009 2.009.899 2.009 2.009 2.009c.555 0 1.057-.225 1.42-.588.363-.353.589-.846.589-1.391 0-.011 0-.021 0-.032v.002zm-11.125.01c0 1.602-1.299 2.901-2.901 2.901s-2.901-1.299-2.901-2.901c0-.012 0-.026 0-.041 0-.788.328-1.499.854-2.005l.001-.001c.51-.521 1.221-.844 2.007-.844h.048-.002.041c.785 0 1.494.324 2.001.846l.001.001c.524.508.85 1.218.85 2.005v.044-.002zm12.02-.01v.004c0 1.602-1.299 2.901-2.901 2.901s-2.901-1.299-2.901-2.901 1.299-2.901 2.901-2.901c.801 0 1.527.325 2.052.85.524.508.849 1.218.849 2.004v.045zm-10.041.031c0-.021.001-.045.001-.07 0-1.307-.541-2.487-1.41-3.33l-.001-.001c-.844-.871-2.024-1.412-3.331-1.412-.025 0-.049 0-.074.001h.004c-.01 0-.021 0-.032 0-.875 0-1.694.24-2.394.658l.021-.012c-1.443.85-2.397 2.397-2.397 4.166s.953 3.316 2.374 4.154l.022.012c.679.406 1.498.646 2.373.646h.036-.002c.021 0 .045.001.07.001 1.307 0 2.487-.541 3.33-1.41l.001-.001c.871-.844 1.412-2.024 1.412-3.331 0-.024 0-.049-.001-.073v.004zm6.96-5.969c-1.66-.729-3.594-1.154-5.628-1.154-.057 0-.115 0-.172.001h.009c-.071-.001-.154-.002-.238-.002-2.068 0-4.038.422-5.827 1.184l.097-.037h.017c.832 0 1.624.175 2.34.489l-.037-.015c.739.315 1.37.746 1.901 1.276.526.534.955 1.167 1.255 1.867l.015.039c.3.679.474 1.471.474 2.303v.018-.001c0-1.619.641-3.088 1.682-4.168l-.002.002c.519-.524 1.131-.955 1.811-1.264l.038-.016c.662-.311 1.436-.501 2.252-.521h.007zm4.99 5.968c-.003-2.653-2.154-4.802-4.807-4.802-2.655 0-4.807 2.152-4.807 4.807s2.152 4.807 4.807 4.807c1.329 0 2.531-.539 3.402-1.41.868-.844 1.406-2.023 1.406-3.327 0-.026 0-.053-.001-.079zm-2.77-5.895h3.989c-.298.346-.558.738-.766 1.16l-.014.032c-.181.344-.325.743-.412 1.163l-.005.03c.715.972 1.144 2.193 1.144 3.515 0 3.297-2.673 5.97-5.97 5.97-1.864 0-3.529-.855-4.624-2.194l-.008-.011c-.436.541-.879 1.15-1.291 1.781l-.054.087c-.191-.333-.376-.613-.576-.882l.016.023q-.442-.63-.776-1.015c-1.103 1.324-2.753 2.161-4.598 2.161-3.298 0-5.971-2.673-5.971-5.971 0-1.298.414-2.499 1.117-3.478l-.012.018c-.092-.45-.236-.849-.429-1.218l.012.025c-.223-.455-.484-.847-.789-1.2l.006.007h3.802c1.062-.701 2.293-1.26 3.61-1.611l.088-.02c1.327-.373 2.851-.588 4.425-.588h.079-.004.107c1.523 0 2.995.213 4.39.611l-.113-.027c1.378.383 2.584.942 3.672 1.665l-.047-.029z"/>
    </SVG>
);

const tripadvisorSocialLink = {
	name: 'tripadvisor',
	attributes: { service: 'tripadvisor' },
	title: 'Trip Advisor',
	icon: tripadvisorIcon,
};

wp.hooks.addFilter(
	'blocks.registerBlockType',
	'tripadvisor-social-icon',
	function( settings, name ) {
		if ( 'core/social-link' === name ) {
			if ( settings.variations.length ) {
				var variation = tripadvisorSocialLink;
				if ( ! variation.isActive ) {
					variation.isActive = ( blockAttributes, variationAttributes ) => blockAttributes.service === variationAttributes.service;
				}
				settings.variations.push( variation );
			}
		}
		return settings;
	}
);

wp.domReady( function() {
	wp.blocks.registerBlockVariation( 'core/social-link', tripadvisorSocialLink );
});
```

hope it helps to someone else too.

mkkeck commented:
> > 
> 
> I'm sorry to ask @mkkeck, but I can't get it working. Could you share your full plugin with social-services.json file by any change? Sorry for asking!

I've updated the ghist:
https://gist.github.com/mkkeck/84fe5fda939002fcf210d66b99de7c0f

Simply copy the code and put it into a new file (php class) under your child theme. 

Created a json file with all your icons and save it as social-services.json under your child theme:

Example json data:

```
{
  "socialServices": [
     "share": {
       "name": "Share",
        "icon": "<svg ...><path d='...' /></svg>",  
        "keywords": ["optional", "key", "words"]
       }
  ]
}
```

gaambo commented:
I want to add my voice to this. There will always be social platforms / links that core won't provide - either because they are niche-platforms or because they are very regional/local. Another example: I like to use the social links block to add an icon to download the app and use a custom icon + link there. 

@ndiego social sharing link is actually pretty similar to the core social-link one, and he introduced custom icons in a recent commit (https://github.com/ndiego/social-sharing-block/pull/23).

I think the way this should work is pretty simple: 
- [x] In the editor, allow extenders to register custom variations for the core/social-link block with title + name. Already works in core.
- [ ] Use the active variations icon + title in the editor to display custom icons/titles
- [ ] Add a PHP filter to allow adding custom services to the output

I don't think the social icons block needs to be a full-fledged icons block (although in the future, that is absolutely interesting). But I think as a first step, using the variations and adding a PHP filter is a quick fix for this. And something that should have been in there since the beginning - unfortunately, a lot of core blocks are not easily extendable.

Before I saw this PR I even worked on [my own PR for this issue](https://github.com/WordPress/gutenberg/pull/59368), based on the code from the social-sharing-block. You can find it here: 
It's similar to this one, but uses the variations' data directly. 

What are the blockers here atm? Is this a discussion about registering custom icons or are there concrete bugs with the implementation in the React Native side?

tomjn commented:
Some pragmatic thoughts:

 - backwards compatibility means once a service is added to the core block, removing it or even changing its icon is a non-trivial task
     - this means that the intended aims of being able to interchange global styles and templates from multiple themes would be broken by this
 - letting users add their own via code means the block would break when copied to another site, into a pattern, or when the theme is changed (most users mistakenly add this code via `functions.php`)
 - Some users will abuse this to do things it was never intended for that are semantically bad because it looks similar to what they want (e.g. the person who mentioned here that they use it for app download links)
   - This has consequences for any attempts to use semantic markup to indicate links to social profiles leading to unintended side effects, especially if it's picked up by other things
   - Would a generic icon links block have been a better fit?
 - **when a block references a custom social network that the core block does not support the behaviour is undefined**

Note that this doesn't mean I do not support this, just that there are practical hurdles to overcome that aren't being discussed as much as they should. All of these are issues that need to be solved.



> Use the active variations icon + title in the editor to display custom icons/titles

My understanding of block variations is that this is already the case, when defining your variation you can specify its icon. This is how all the `core/embed` blocks currently work

gaambo commented:
I even think we could do without the custom/new filter on rendering, if instead the variations are registered on the server and used for rendering as well. 
With the new `get_block_type_variations` filter, [we kind of have a way to register custom variations on the server side](https://github.com/WordPress/gutenberg/issues/47170#issuecomment-1906496314). So the PHP rendering could just be as the editor rendering, by using the attributes/icons provided by the variation.

creative-andrew commented:
What is the status on this?

Mamaduka commented:
Closing this in favor of #59368.
